mart/callbacks/adversarial_training.py

```diff
@@ -34,35 +48,54 @@ def setup(self, trainer, pl_module, stage=None):
     def teardown(self, trainer, pl_module, stage=None):
         pl_module.on_after_batch_transfer = self._on_after_batch_transfer
 
+    def wrap_model(self, model, dataloader_idx):
+        """Make a model, such that `output = model(batch)`."""
+
+        # Consume dataloader_idx
+        if hasattr(model, "attack_step"):
+
+            def model_forward(batch):
+                output = model.attack_step(batch, dataloader_idx)
+                return output
+
+        elif hasattr(model, "training_step"):
+            # Monkey-patch model.log to avoid spamming.
+            def model_forward(batch):
+                with MonkeyPatch(model, "log", lambda *args, **kwargs: None):
+                    output = model.training_step(batch, dataloader_idx)
+                return output
+
+        else:
+            model_forward = model
+
+        return model_forward
+
     def on_after_batch_transfer(self, pl_module, batch, dataloader_idx):
         batch = self._on_after_batch_transfer(batch, dataloader_idx)
 
-        # FIXME: Remove use of step
+        adversary = None
+
         trainer = pl_module.trainer
         if trainer.training:
             adversary = self.train_adversary
-            step = "training"
         elif trainer.validating:
             adversary = self.validation_adversary
-            step = "validation"
         elif trainer.testing:
             adversary = self.test_adversary
-            step = "test"
-        else:
-            return batch
 
-        # Create attacked model where the adversary executes before the model
-        # FIXME: Should we just use pl_module.training_step? Ideally we would not decompose batch
-        #        and instead pass batch directly to the underlying pl_module since it knows how to
-        #        interpret batch.
-        def attacked_model(input, **batch):
-            input_adv = adversary(input=input, **batch)
-            return pl_module(input=input_adv, **batch)
+        # Skip if adversary is not defined for all phases train/validation/test.
+        if adversary is None:
+            return batch
 
         # Move adversary to same device as pl_module and run attack
-        # FIXME: Directly pass batch instead of assuming it has a structure?
-        input, target = batch
         adversary.to(pl_module.device)
-        input_adv = adversary(input=input, target=target, step=step, model=attacked_model)
 
-        return [input_adv, target]
+        # We assume Adversary is not aware of PyTorch Lightning,
+        # so wrap the model as `output=model(batch)`.
+        model = self.wrap_model(pl_module, dataloader_idx)
```

Comment:
Why is this not done in `Adversary`?

Reply:
You are right. I think we should have something like `model_transform` in `Adversary` that is configurable.

Reply:
I removed the wrapper in callback and added a configurable model_transform in `Adversary` in 92f9fe23fcb41c876222ce12881e59f9133067cb